### PR TITLE
Platform Specifics

### PR DESCRIPTION
**Pull request by samhouts**
### Description of Change ###

Introducing Platform Specifics! Features or behaviors that apply to one platform only can now be implemented without requiring custom renderers. These new features/behaviors can then be accessed easily via a fluent code API or XAML.

Vendors can easily add their own Platform Specifics by attaching Effects to them (see 63a924d77e10a4d637b1b2c2709f4613f4a4e532 and https://github.com/xamarin/Xamarin.Forms/commit/1f9482e4b5969f19fcde5c691443fcccc848c54e for complete example).

This PR implements the framework that enables the new API and also includes several examples of platform specific features, which can be previewed using the Platform Specifics gallery page:

- Blur support for any `VisualElement` on iOS
- Translucent navigation bar on iOS
- Partially collapsed navigation bar (with icons!) on `MasterDetailPage` on Windows
- Toolbar placement options on Windows
- AdjustResize/AdjustPan on Android (known issue: AdjustResize disables status bar color)

### Bugs Fixed ###

- [Bug 37202  - [Android] Content inside ScrollView not scrollable] (https://bugzilla.xamarin.com/show_bug.cgi?id=37202)
- [Bug 39531 - [UWP] ToolbarItems Displayed at the Top of the Page on Mobile Devices] (https://bugzilla.xamarin.com/show_bug.cgi?id=39531)
- [Bug 39765  - WindowSoftInputMode Attribute is Ignored When Using AppCompat] (https://bugzilla.xamarin.com/show_bug.cgi?id=39765)
- [Bug 39916  - SoftKeyboard default behavior covers Listview when using FormsAppCompatActivity.] (https://bugzilla.xamarin.com/show_bug.cgi?id=39916)

### API Changes ###

Added Namespaces:
- `Xamarin.Forms.PlatformConfiguration.AndroidSpecific`
- `Xamarin.Forms.PlatformConfiguration.iOSSpecific`
- `Xamarin.Forms.PlatformConfiguration.WindowsSpecific`
- `Xamarin.Forms.PlatformConfiguration`

Added method to all controls + `Xamarin.Forms.Application`:
- `IPlatformElementConfiguration<TElement> On<TPlatform>();`

Added method to `Element`:
- `bool EffectIsAttached(string name);`

Added many new types for API & sample features. See af73febbf8b5c2236ce8a19b760d6092cebd3249  for details.

### Behavioral Changes ###

None, unless opt in to new features 

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [x] Consolidate commits as makes sense

### CODE REVIEW

File: Xamarin.Forms.Core/PlatformConfiguration/WindowsSpecific/ToolbarPlacement.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
+namespace Xamarin.Forms.PlatformConfiguration.WindowsSpecific
+{
+	public enum ToolbarPlacement
+	{
+		/// <summary>
+		/// Default will place the Toolbar at the bottom of the screen on phones and
+		/// at the top of the screen on larger devices
+		/// </summary>
+		Default,
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/PlatformConfiguration/WindowsSpecific/ToolbarPlacement.cs]: reminder: we don't fo inline docs

**Comment by StephaneDelcroix** [comment]
don't we have any other ways than adding the platformspecifics in the Core ? that makes it harder to further expand, imho

File: Xamarin.Forms.Core/PlatformConfiguration/WindowsSpecific/ToolbarPlacement.cs (hunk 1 above)
Comment (hartez) [inline Xamarin.Forms.Core/PlatformConfiguration/WindowsSpecific/ToolbarPlacement.cs]: Why is that?

**Comment by hartez** [comment]
@StephaneDelcroix I suppose we could split them out into a separate project/assembly (or separate projects/assemblies for each platform). What situation are you envisioning where having them in the core causes a problem?

File: Xamarin.Forms.Core/PlatformConfigurationRegistry.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
+using System;
+using System.Collections.Generic;
+
+namespace Xamarin.Forms
+{
+	/// <summary>
+	/// Helper that handles storing and lookup of platform specifics implementations
+	/// </summary>
+	/// <typeparam name="TElement">The Element type</typeparam>
+	internal class PlatformConfigurationRegistry<TElement> : IElementConfiguration<TElement>
```
Comment (jassmith) [inline Xamarin.Forms.Core/PlatformConfigurationRegistry.cs]: We could lazily create these classes it looks like rather than create them eagerly.

**Comment by StephaneDelcroix** [comment]
@hartez adding a n+1 plat without modifying the core

File: Xamarin.Forms.Core/PlatformConfigurationRegistry.cs (hunk 2 above)
Comment (hartez) [inline Xamarin.Forms.Core/PlatformConfigurationRegistry.cs]: Done.

**Pull request merged**